plugins/curation/server/__init__.py

```diff
@@ -0,0 +1,198 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+###############################################################################
+#  Copyright 2016 Kitware Inc.
+#
+#  Licensed under the Apache License, Version 2.0 ( the "License" );
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+###############################################################################
+
+from girder.api import access
+from girder.api.describe import Description, describeRoute
+from girder.api.rest import Resource, loadmodel
+from girder.constants import AccessType
+from girder.utility import mail_utils
+import datetime
+import posixpath
+
+
+CURATION = 'curation'
+
+ENABLED = 'enabled'
+STATUS = 'status'
+TIMELINE = 'timeline'
+ENABLE_USER_ID = 'enableUserId'
+REQUEST_USER_ID = 'requestUserId'
+REVIEW_USER_ID = 'reviewUserId'
+
+CONSTRUCTION = 'construction'
+REQUESTED = 'requested'
+APPROVED = 'approved'
+
+DEFAULTS = {
+    ENABLED: False,
+    STATUS: CONSTRUCTION,
+}
+
+
+class CuratedFolder(Resource):
+
+    @access.public
+    @loadmodel(model='folder', level=AccessType.READ)
+    @describeRoute(
+        Description('Get curation details for the folder.')
+        .param('id', 'The folder ID', paramType='path')
+        .errorResponse('ID was invalid.')
+        .errorResponse('Read permission denied on the folder.', 403)
+    )
+    def getCuration(self, folder, params):
+        result = dict(DEFAULTS)
+        result[TIMELINE] = []
+        result.update(folder.get(CURATION, {}))
+        return result
+
+    @access.public
+    @loadmodel(model='folder', level=AccessType.READ)
+    @describeRoute(
+        Description('Set curation details for the folder.')
+        .param('id', 'The folder ID', paramType='path')
+        .param('enabled', 'Enable or disable folder curation.',
+               required=False, dataType='boolean')
+        .param('status', 'Set the folder curation status.',
+               required=False, dataType='string')
+        .errorResponse('ID was invalid.')
+        .errorResponse('Write permission denied on the folder.', 403)
+    )
+    def setCuration(self, folder, params):
```

Comment:
There is a lot of boolean logic in this endpoint that was hard for me to reason about, as I needed to think about enabled/disabled + state transitions + user perms.

I'd prefer to see this broken into  separate endpoints, and we should also be able to rely on the decorators to enforce permissions in most (all?) of the cases as well as getting more parameter validation.

The following is just a suggestion, feel free to go with a different design.  I think it simplifies reasoning, especially about permissions --which I'd rather not do, so as to minimize access bugs, and because I'm not good at it. 🐒 

     @access.user(scope=TokenScope.DATA_OWN)
     @loadmodel(model='folder', level=AccessType.OWN)
     @describeRoute(
         Description('Enable or disable curation for the folder.')
         .param('id', 'The folder ID', paramType='path')
         .param('enabled', 'Enable or disable folder curation.',
                required=True, dataType='boolean')
         .errorResponse('ID was invalid.')
         .errorResponse('Admin permission denied on the folder.', 403)
     )
     def setCuration(self, folder, params):
     ...

     @access.user(scope=TokenScope.DATA_WRITE)
     @loadmodel(model='folder', level=AccessType.WRITE)
     @describeRoute(
         Description('Change the curation state from construction to requested for the folder.')
         .param('id', 'The folder ID', paramType='path')
         .errorResponse('ID was invalid.')
         .errorResponse('Write permission denied on the folder.', 403)
     )
     def setCurationRequested(self, folder, params):
         raise something if not curationEnabled(folder) 
         ...

     @access.user(scope=TokenScope.DATA_OWN)
     @loadmodel(model='folder', level=AccessType.OWN)
     @describeRoute(
         Description('Change the curation state for the folder.')
         .param('id', 'The folder ID', paramType='path')
         .param('status', 'Set the folder curation status.',
                required=True, dataType='string')
         .errorResponse('ID was invalid.')
         .errorResponse('Admin permission denied on the folder.', 403)
     )
     def setCurationState(self, folder, params):
         raise something if not curationEnabled(folder) 
         ...